osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs

```diff
@@ -1,25 +1,80 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using osu.Game.Beatmaps;
 using osu.Game.Rulesets.Judgements;
-using osu.Game.Rulesets.Objects;
 using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Judgements;
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Rulesets.Osu.Scoring
 {
     public partial class OsuHealthProcessor : DrainingHealthProcessor
     {
+        private ComboResult currentComboResult = ComboResult.Perfect;
+
         public OsuHealthProcessor(double drainStartTime, double drainLenience = 0)
             : base(drainStartTime, drainLenience)
         {
         }
 
-        protected override int? GetDensityGroup(HitObject hitObject) => (hitObject as IHasComboInformation)?.ComboIndex;
-
         protected override double GetHealthIncreaseFor(JudgementResult result)
+        {
+            if (IsSimulating)
+                return getHealthIncreaseFor(result);
+
+            if (result.HitObject is not IHasComboInformation combo)
+                return getHealthIncreaseFor(result);
+
+            if (combo.NewCombo)
+                currentComboResult = ComboResult.Perfect;
+
+            switch (result.Type)
+            {
+                case HitResult.LargeTickMiss:
+                case HitResult.Ok:
+                    setComboResult(ComboResult.Good);
+                    break;
+
+                case HitResult.Meh:
+                case HitResult.Miss:
+                    setComboResult(ComboResult.None);
+                    break;
+            }
+
+            // The tail has a special IgnoreMiss judgement
```

Comment:
...or `SmallTickMiss`, if classic mod is active. But the condition below appears to cover both, so it's fine.